ClassicLootManager/Modules/Rosters/RosterManager/ExpressionParser.lua

```diff
@@ -0,0 +1,425 @@
+                errors[#errors + 1] = errorMessage
+                return
+            end
+        end
+    elseif operationType == OPERATION_TYPE.AND then
+        for _, pattern in ipairs(patterns) do
+            if not expression:match(pattern) then
+                return
+            end
+        end
+        errors[#errors + 1] = errorMessage
+    else
+        error("Invalid operation type: " .. operationType)
+    end
+end
+
+local function verifyConditions(expression, conditions, errors)
+    for _, condition in ipairs(conditions) do
+        verifyPattern(expression, condition[1], condition[2], condition[3], errors)
+    end
+end
+
+local function verifyAllTextIsVariableNames(expression, variables, errors)
+    local errorMessage = CLM.L["Expression may only contain the pre-defined variables"]
+
+    local variableValues = {}
+    for _, value in pairs(variables) do
+        variableValues[value] = true
+    end
+
+    local nonOperatorParenthesisOrNumber = "[^%+%-%/%*%^%(%)%d%.]+"
+    for word in expression:gmatch(nonOperatorParenthesisOrNumber) do
+        if not variableValues[word] then
+            errors[#errors + 1] = errorMessage
+            return
+        end
+    end
+end
+
+local function verifyParenthesisAreBalanced(expression, errors)
+    local errorMessage = CLM.L["Expression has unbalanced parentheses"]
+
+    local nestDepth = 0
+    for i = 1, #expression do
+        local c = expression:sub(i, i)
+        if c == '(' then
+            nestDepth = nestDepth + 1
+        elseif c == ')' then
+            nestDepth = nestDepth - 1
+        end
+        if nestDepth < 0 then
+            errors[#errors + 1] = errorMessage
+            return
+        end
+    end
+    if nestDepth ~= 0 then
+        errors[#errors + 1] = errorMessage
+    end
+end
+
+local function tokenize(input)
+    local tokens = {}
+    local currentToken = ""
+    local index = 1;
+    for char in input:gmatch(".") do
+        if char:match("[%d%.]") or (char:match("-") and not currentToken:match("[%d]$")) then
+            -- gathering token parts
+            currentToken = currentToken .. char
+        else
+            -- write current token
+            currentToken = currentToken:gsub("%-%-", "") -- correction for negative numbers
+            tokens[index] = currentToken
+            currentToken = ""
+            index = index + 1
+
+            -- write new token
+            tokens[index] = char
+            index = index + 1
+        end
+    end
+    if currentToken ~= "" then
+        tokens[index] = currentToken
+    end
+    return tokens
+end
+
+local function processOperator(tokens, process)
+    for i, token in ipairs(tokens) do
+        if process[token] then
+            local j = i - 1
+            while tokens[j] == "" do
+                j = j - 1
+            end
+            local left = tonumber(tokens[j])
+            local right = tonumber(tokens[i + 1])
+            local result = process[token](left, right)
```

Comment:
be sure to check if `process[token]` exists before calling

Reply:
I dont think it should be possible with this setup for it to not exist. I can add a check if you like, but could you give an example?